src/middleware/combine.ts

```diff
@@ -1,25 +1,60 @@
-import { GetState, SetState, State, StoreApi } from '../vanilla'
-import { NamedSet } from './devtools'
-
-type Combine<T, U> = Omit<T, keyof U> & U
-
-export const combine =
-  <PrimaryState extends State, SecondaryState extends State>(
-    initialState: PrimaryState,
-    create: (
-      // Note: NamedSet added for convenience
-      set: SetState<PrimaryState> & NamedSet<PrimaryState>,
-      get: GetState<PrimaryState>,
-      api: StoreApi<PrimaryState>
-    ) => SecondaryState
-  ) =>
-  (
-    set: SetState<Combine<PrimaryState, SecondaryState>>,
-    get: GetState<Combine<PrimaryState, SecondaryState>>,
-    api: StoreApi<Combine<PrimaryState, SecondaryState>>
-  ) =>
-    Object.assign(
-      {},
-      initialState,
-      create(set as any, get as any, api as any)
-    ) as Combine<PrimaryState, SecondaryState>
+import { Store, UnknownState, StoreInitializer } from '../vanilla'
+
+// ============================================================================
+// Types
+
+type Combine =
+  <T extends UnknownState, U extends UnknownState>
+    ( initialState: T
+    , additionalStateCreator:
+        ( set: Store<O.Overwrite<T, U>>['setState']
+        , get: Store<O.Overwrite<T, U>>['getState']
+        , store: Store<O.Overwrite<T, U>>
+        ) => U
+    ) =>
+      StoreInitializer<O.Overwrite<T, U>, Store<O.Overwrite<T, U>>>
```

Comment:
Previously the combine types were something like this...
```ts
type Combine = 
  <T, U>
    ( controlled: T
    , createController:
        ( set: (state: Partial<T>, replace?: boolean) => void
        , get: () => T
        )
          => U
    ) =>
        ( set: (state: Partial<Overwrite<T, U>, replace?: boolean) => T
        , get: () => Overwrite<T, U>
        )
          => Overwrite<T, U>

type Overwrite<T, U> = Omit<T, keyof U> & U
```

Now these types are lying, the actual types are these...
```diff
  type Combine = 
    <T, U>
      ( controlled: T
      , createController:
-         ( set: (state: Partial<T>, replace?: boolean) => void
+         ( set: (state: Partial<Overwrite<T, U>>, replace?: boolean) => void
-         , get: () => T
+         , get: () => Overwrite<T, U>
          )
            => U
      ) =>
          ( set: (state: Partial<Overwrite<T, U>, replace?: boolean) => T
          , get: () => Overwrite<T, U>
          )
            => Overwrite<T, U>

type Overwrite<T, U> = Omit<T, keyof U> & U
```
...how do I know this? Because that is what the runtime behavior is, `get` would return `Overwrite<T, U>` and not `T` and it matters because I can create many runtime bugs that would compile. Examples...

```ts
let store = create(combine(
  {
    foo: 1,
    bar: 2
  },
  (set, get) => ({
    getSum: () => values(get()).reduce((a, b) => a + b, 0)
  })
))

const values =
  Object.values as (<T>(t: T) => T[keyof T][])

// technically this type of `Object.values` is incorrect
// but still usually people do it when they know what they're doing
// so one could write this thinking get only returns `T`

console.log(store.getState().getSum())
```

Now this compiles because `get()` says it'll return `{ a, b }` but during runtime it'll return `{ a, b, getSum }` and it'll [log `3() => values(get()).reduce((a, b) => a + b, 0)` instead of `3`](https://stackblitz.com/edit/typescript-mvcews?file=tsconfig.json).

This is just one example, it in general sets up a wrong precedence that `set` and `get` only operate with `T` whereas they operate with the whole store ie `Overwrite<T, U>`. Another example is doing `set(foo, true)` thinking it'll only replace the "controlled" `T` state whereas in reality it'll replace the whole state. I can make many more scenarios but you get the point.

So rewrite makes sure we're not lying in the types.

Reply:
Fixing types without changing the behavior is nice.